### ISSUE TITLE
How to properly layout a lib and main in the same process and letting both use the same sub modules?

### ISSUE THREAD

**Issue by neilyoung** [rust-lang/rust#78266]
The entire problem is outlined here https://users.rust-lang.org/t/confusion-about-module-name-ambiguity/50512 w/o having found a satisfying solution. 

The core of the problem has been summarized by a user in the thread, so I quote that here:

Project is `cargo new lib-test``

# This works

```
// src/main.rs
mod lib;
fn main {}
```
```
// src/lib.rs
```

# This doesn't compile

```
// src/main.rs
mod lib;
fn main {}
```

```
// src/lib.rs
mod test;
```

```
// src/test.rs
```

The error suggests help: to create the module `test`, create file "src/lib/test.rs". Doing this gives the reverse suggestion: create src/test.rs

The thread suggests to remove `mod lib;` from `main.rs` and call the mod test functions provided by `lib.rs`:

```
// src/test.rs
pub fn test()  {
}
```

```
// src/lib.rs
pub mod test;

pub fn test() {}
```

```
use lib_test::test::test;

fn main() {
    test();
}

```

`lib.rs` needs to be altered so, that `pub mod test;` is exported.

This works, compiles w/o warning and runs. It just makes VSCode to give an error notation in `lib.rs`:

![image](https://user-images.githubusercontent.com/731020/96973687-86ce5680-1518-11eb-89d8-c815df4c840d.png)

So the question is: Is the suggested solution the "right" solution to do this task? Calling functions from one and the same modules in a lib _and_ a bin?

**Comment by neilyoung** [comment]
VSCode problem gone after restart of VSCode (most likely the restart of the Rust Analyzer would have been sufficient)

**Comment by Heliozoa** [comment]
As I mentioned in the thread, I think it would be preferable for Cargo to outright forbid using a crate root file as a module. The current behavior is confusing and surprising, and the related error messages are unhelpful. That said I'm not familiar with how Cargo actually works, so there may be some issues with this suggestion.

The module system is a common point of confusion among newer users on URLO even without this stumbling block, though I think it's good once you do understand it. [This thread](https://users.rust-lang.org/t/n00b-question-how-do-i-let-benchmark-files-know-my-functions/50509) was posted the same day, with the OP similarly trying `mod lib;` with no success. I've done the same mistake myself in the past, and I feel that it's an easy mistake to make. Given all this I think it would be great if this was addressed in Cargo in some way.

**Comment by jonas-schievink** [comment]
The user forum is the right place to ask questions like this, we use this issue tracker only for bugs and small, concrete enhancement requests. If you have a specific, actionable enhancement request for Cargo, please file it at https://github.com/rust-lang/cargo, since this repo is for the Rust compiler and standard library.

**Comment by neilyoung** [comment]
Well, I did not ask for much really. Just a short confirmation that the approach discussed in the forum is correct and that I'm not going to fly 10 miles down the road.  Especially for beginners in the language a correct "Separation of Concerns" is hardly possible.  Is it rust? Is it rustc? Is it even cargo?

A friendly hint would have been enough for me, where I already tried to discuss the problem at the "right" place and prepare it bite-sized. And since I'm not the only one, who was trapped, it would have been helpful for others too.

Thanks anyway to @Heliozoa , who obviously liked this answer. Not me, honestly.
